src/spatialdata_plot/pl/render.py

```diff
@@ -213,26 +230,64 @@ def _render_shapes(
             else None
         )
 
-        ds_result = (
-            ds.tf.shade(
+        if color_by_categorical or col_for_color is None:
+            ds_cmap = None
+            if color_vector is not None:
+                ds_cmap = color_vector[0]
+                if isinstance(ds_cmap, str) and ds_cmap[0] == "#":
+                    ds_cmap = ds_cmap[:-2]
+
+            ds_result = ds.tf.shade(
                 agg,
-                cmap=color_vector[0][:-2],
+                cmap=ds_cmap,
                 color_key=color_key,
-                min_alpha=np.min([150, render_params.fill_alpha * 255]),
+                min_alpha=np.min([254, render_params.fill_alpha * 255]),
+                how="linear",
             )
-            if color_by_categorical or col_for_color is None
-            else ds.tf.shade(
+        elif aggregate_with_reduction is not None:  # to shut up mypy
+            ds_cmap = render_params.cmap_params.cmap
+            # in case all elements have the same value X: we render them using cmap(0.0),
+            # using an artificial "span" of [X, X + 1] for the color bar
+            # else: all elements would get alpha=0 and the color bar would have a weird range
+            if aggregate_with_reduction[0] == aggregate_with_reduction[1]:
+                ds_cmap = matplotlib.colors.to_hex(render_params.cmap_params.cmap(0.0), keep_alpha=False)
+                aggregate_with_reduction = (aggregate_with_reduction[0], aggregate_with_reduction[0] + 1)
+
+            ds_result = ds.tf.shade(
                 agg,
-                cmap=render_params.cmap_params.cmap,
+                cmap=ds_cmap,
+                how="linear",
+                min_alpha=np.min([254, render_params.fill_alpha * 255]),
+            )
+
+        # shade outlines if needed
+        outline_color = render_params.outline_params.outline_color
+        if isinstance(outline_color, str) and outline_color.startswith("#") and len(outline_color) == 9:
+            outline_color = outline_color[:-2]
```

Comment:
What if the outline color is given as hex without alpha?

Reply:
Then the string would have length 7 and therefore the last 2 characters wouldn't be removed. When I tested it, that worked fine :)